File: 백준/Gold/2239. 스도쿠/스도쿠.py
import sys

input = sys.stdin.readline


def dfs(count):
    if count == len(point):
        for i in range(9):
            print("".join(list(map(str, sudoku[i]))))
        exit(0)
    for i in range(1, 10):
        if (
            hor(i, point[count][1])
            and ver(i, point[count][0])
            and square(i, point[count][0], point[count][1])
        ):
            sudoku[point[count][1]][point[count][0]] = i
            dfs(count + 1)
            sudoku[point[count][1]][point[count][0]] = 0


def hor(num, y):
    for i in range(9):
        if sudoku[y][i] == num:
            return False
    return True


def ver(num, x):
    for i in range(9):
        if sudoku[i][x] == num:
            return False
    return True


def square(num, x, y):
    for i in range(3 * (y // 3), 3 * (y // 3) + 3):
        for j in range(3 * (x // 3), 3 * (x // 3) + 3):
            if sudoku[i][j] == num:
                return False
    return True


sudoku = [list(map(int, input().strip())) for _ in range(9)]
point = []
for i in range(9):
    for j in range(9):
        if sudoku[i][j] == 0:
            point.append([j, i])
dfs(0)

File: 백준/Gold/2239. 스도쿠/스도쿠.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
//	static StringTokenizer st;
	static int[][] g = new int[9][9];
	static boolean[][] row = new boolean[9][10];
	static boolean[][] col = new boolean[9][10];
	static boolean[][] square = new boolean[9][10];

	public static void main(String[] args) throws Exception {
		for (int i = 0; i < 9; i++) {
			String s = br.readLine();
			for (int j = 0; j < 9; j++) {
				int num = s.charAt(j) - '0';
				g[i][j] = num;
				if (num != 0) {
					row[i][num] = true;
					col[j][num] = true;
					square[(i / 3) * 3 + (j / 3)][num] = true;
				}
			}
		}
		dfs(0);
	}

	static void dfs(int d) {
		if (d == 81) {
			for (int i = 0; i < 9; i++) {
				for (int j = 0; j < 9; j++) {
					sb.append(g[i][j]);
				}
				sb.append("\n");
			}
			System.out.println(sb);
			System.exit(0);
		}
		int x = d / 9;
		int y = d % 9;
		if (g[x][y] != 0) {
			dfs(d + 1);
			return;
		}
		int block = (x / 3) * 3 + (y / 3);
		for (int i = 1; i <= 9; i++) {
			if (!row[x][i] && !col[y][i] && !square[block][i]) {
				g[x][y] = i;
				row[x][i] = true;
				col[y][i] = true;
				square[block][i] = true;
				dfs(d + 1);
				g[x][y] = 0;
				row[x][i] = false;
				col[y][i] = false;
				square[block][i] = false;
			}
		}
	}
}